File: Task1/Indexer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Task1
{
    public class Indexer
    {
        double[] array;
        int firstIndex, lastIndex;
        public Indexer(double[] array, int firstIndex, int length)
        {
            if (firstIndex < 0 || length <= 0 || array.Length <= firstIndex + length)
                throw new ArgumentException();
            this.array = array;
            this.firstIndex = firstIndex;
            this.lastIndex = firstIndex + length - 1;
        }
        public int Length
        {
            get => this.lastIndex - this.firstIndex + 1;
        }
        public double this[int index]
        {
            get
            {
                if (index < 0 || index >= this.Length)
                    throw new IndexOutOfRangeException();
                return this.array[index + firstIndex];
            }
            set
            {
                if (index < 0 || index >= this.Length)
                    throw new IndexOutOfRangeException();
                this.array[index + firstIndex] = value;
            }
        }

    }
}

File: Task1Cpp/Indexer.cpp
#include "Indexer.h"
#include <stdexcept>

Indexer::Indexer(double* array, int arrayLength, int firstIndex, int length) {
	if (firstIndex < 0 || length <= 0 || firstIndex + length >= arrayLength)
		throw std::invalid_argument("Incorrect first index or length");
	this->array = array;
	this->firstIndex = firstIndex;
	this->lastIndex = firstIndex + length - 1;
}

int Indexer::Length() {
	return this->lastIndex - this->firstIndex + 1;
}

double& Indexer::operator[] (const int index) {
	return this->array[index + firstIndex];
}